cafebabel/static/js/edit-article.js

```diff
@@ -27,68 +39,108 @@ class Tags {
 
-function addTagsRemoveListener() {
-  const tagsButtonRemove = document.querySelectorAll('.tags .tags-list')
-  tagsButtonRemove.forEach(tagButtonRemove =>
+    tagButtonAdd.addEventListener('click', event => {
+      event.preventDefault()
+      const submission = event.target.previousSibling.value
+      const tags = new Tags()
+      if (tags.checkSubmission(submission)) tags.add(submission)
+      tags.display()
+      Effect.fadeIn(document.querySelector('.tags .tags-list li:last-child'))
+      document.querySelector('.tags input[name=tag-new]').value = ''
+    })
+  }
+  static clickRemove(tagButtonRemove) {
     tagButtonRemove.addEventListener('click', event => {
-      const tags = new Tags(document.querySelector('.tags'))
-      tags.remove(event.target.innerText)
-      displayTags(tags)
+      const tags = new Tags()
+      const tagElement = event.target.parentNode
+      tags.remove(tagElement.innerText)
+      Effect.fadeOut(tagElement).then(response => tags.display())
     })
-  )
-}
+  }
 
-window.addEventListener('load', addTagsRemoveListener)
+  static keyup() {
+    function handleSuggestion(tag) {
+      const tagsSuggestion = document.querySelector('#tags-suggestions')
+      tagsSuggestion.innerHTML = ''
+      const li = document.createElement('li')
+      li.append(tag.name)
+      tagsSuggestion.appendChild(li)
+      tagsSuggestion.classList.remove('inactive')
+      tagsSuggestion.addEventListener('click', event => {
+        const submission = event.target.innerText
+        const tags = new Tags()
+        if (tags.checkSubmission(submission)) tags.add(submission)
+        tags.display()
+        li.remove()
+        document.querySelector('.tags input[name=tag-new]').value = ''
+        tagsSuggestion.classList.add('inactive')
+      })
+    }
 
-function displayTags(tags) {
-  const container = tags.list.cloneNode(false)
-  tags.list.replaceWith(tags._createTagsList(container, tags.values))
-  document.querySelector('.tags .tags-list li:last-child').classList.add('fadeIn')
-  addTagsRemoveListener()
+    const inputNewTag = document.querySelector('.tags input[name=tag-new]')
+
+    inputNewTag.addEventListener('keyup', event => {
+      /* Return if arrow up, arrow down or enter are pressed */
+      if (event.keyCode == 40 || event.keyCode == 38 || event.keyCode == 13)
+        return
+      const submission = event.target.value
+      const languages = document.querySelector('#language')
+      const language = languages.options[languages.selectedIndex].value
+      if (submission.length < 3) return
+      request(`/article/tag/suggest/?language=${language}&terms=${submission}`)
+        .then(response => response.forEach(handleSuggestion))
+        .catch(console.error.bind(console))
+    })
+  }
 }
+
+Listener.clickAdd()
+Listener.keyup()
+
+window.addEventListener('load', () => {
+  const tags = new Tags(document.querySelector('.tags'))
+  Tags.addTagsRemoveListener()
```

Comment:
This `addTagsRemoveListener` method should probably belong to the currently called `Listener` class?